streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -211,7 +211,15 @@ void initializeStoreOffsetsFromCheckpoint() {
                         log.debug("State store {} initialized from checkpoint with offset {} at changelog {}",
                             store.stateStore.name(), store.offset, store.changelogPartition);
                     } else {
-                        // TODO K9113: for EOS when there's no checkpointed offset, we should treat it as TaskCorrupted
+                        // with EOS, if the previous run did not shutdown gracefully, we would always close
+                        // all tasks as dirty before complete the shutdown which will already wipe out the local
+                        // stores including the checkpoint store. That means, here if the checkpoint file does not
+                        // contain the corresponding offset, there are only two possibilities:
+                        //
+                        // 1. the local state store is also empty, in which case it is safe to restore from scratch.
+                        // 2. the local state store is not empty but it only contains committed records (i.e. only
+                        //    the checkpoint file when corrupted), in which case it is safe to restore and overwrite
+                        //    from scratch too.
```

Comment:
If I understand this reasoning, it's predicated on the previous run doing something during shutdown. I'm not sure we can rely on any such assumptions. For example, what if the previous run got `kill -9`ed instead of requested to shutdown? What if that happened _during_ a normal shutdown? What if the machine crashed, the filesystem got corrupted, and fsck replayed the journal to an arbitrary state during recovery?

I think that we used to delete the checkpoint file as soon as we started processing under EOS, but I can't find that anymore in the code. Is that still the case? If that were true, then this assumption I'm concerned about wouldn't apply, since the checkpoint file wouldn't exist unless it were consistent with the store.

Reply:
That logic is still here, actually now in trunk with or without EOS, we would delete the file right after loading to processor state manager: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java#L227

Reply:
The rationale was that (if you remember :) the only gap where we are at risk of losing all is the commit interval between starting up and the first commit (without EOS we would write that checkpoint file again in the first commit).

Reply:
Thanks @guozhangwang ,

In that case, I'm not understanding the above comment. It says that if the previous shutdown didn't complete gracefully, then we'd delete the store and checkpoint file, but it sounds like in actuality, the checkpoint file wouldn't exist at all in that case, and therefore not be "deleted" on unclean shutdown. In fact, the two conclusions you listed seem to follow from what I just said, so maybe it's just the first part of the comment that I found confusing.

Carry on :)

Reply:
Hmm, that is a good question.

So let's re-think through it:

1) If the state stores are already wiped, then checkpoint file would never exist too.
2) The vice versa, however, is not true: if there's no checkpoint file found or the partition not exist in checkpoint file, the state stores may or may not be empty.

So let's take a closer look at case 2), if the checkpoint file does not exist but the state stores are non-empty, it means 

2.a) we had a clean close before (hence the checkpoint file is written) but then that checkpoint file itself is corrupted or lost. In this case it is safe to just restart restoration from the beginning and overwrite the non-empty stores, since that store was closed and flushed cleanly and hence there's no uncommitted data.

2.b) we had a fatal crash (e.g. kill -9) and the `closeDirty` does not execute at all, and hence there's no checkpoint file but the store may be non-empty. Then upon recovery the fsync may restore to any point in the past. **That's the danger case which we may miss here.**

Reply:
The tricky thing though, is that although by the time when we `initializeOffsetsFromCheckpoint` the store is already initialized, today we do not have a way to check if the `StateStore` is empty or not. But if we just be conservative for 2.b) and always wipe out the store and revive, we would fall into the endless loop.